### PR TITLE
[pvr] if end time is ever less than start time when creating a timer default to instant recording duration

### PR DESCRIPTION
**Pull request by phunkyfish**
## Description

Manual PVR timers do not default to a sensible default if end time is less than start time. This PR uses the instant recording duration to provide that value.

## Motivation and Context

Invalid timer requests could be sent to PVR add-ons.

## How Has This Been Tested?

OSX

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp [hunk 1]
```diff
@@ -587,7 +589,11 @@ void CGUIDialogPVRTimerSettings::Save()
     }
     else if (m_endLocalTime < m_startLocalTime) // Assume the user knows what they are doing, but log a warning just in case
     {
-      CLog::Log(LOGWARNING, "Timer settings dialog: Specified recording end time < start time: expect errors!");
+      CLog::Log(LOGWARNING, "Timer settings dialog: Specified recording end time < start. Setting "
+                            "end time to start time plus instant timer duration minutes.");
```
Comment (ksooo) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:593]: I would log the actual instant timer duration value here.

**Review by ksooo — APPROVED** [review]
As a quick workaround this is okay. Real solution is to detect inconsistencies, to inform the user what's wrong and not to allow timer creation as long as user has not fixed the timer data.

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp [hunk 2]
```diff
@@ -585,9 +587,15 @@ void CGUIDialogPVRTimerSettings::Save()
         }
       }
     }
-    else if (m_endLocalTime < m_startLocalTime) // Assume the user knows what they are doing, but log a warning just in case
+    else if (m_endLocalTime <= m_startLocalTime) // Use a sensible default if end time is invalid
```
Comment (ksooo) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:590]: Why have you removed the `=`? Is start equals end something useful/valid?

**Comment by emveepee** [comment]
Should a manual one-off recording not work like a manual recurring recording? Just replace the day of the week choice with the start time.  There is no error as end time can be less then start time after midnight.   Perhaps a  few moto-cross users will have to record their Enduro's in parts but for most users this UI will be much more intuitive.

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp (hunk 2 above)
Comment (ksooo) [reply xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:590]: Ooops nvm, you inserted the `=`which is fine.

**Comment by phunkyfish** [comment]
> Should a manual one-off recording not work like a manual recurring recording? Just replace the day of the week choice with the start time. There is no error as end time can be less then start time after midnight. Perhaps a few moto-cross users will have to record their Enduro's in parts but for most users this UI will be much more intuitive.

This is no one near that clever.

**Comment by emveepee** [comment]
> 
> 
> > Should a manual one-off recording not work like a manual recurring recording? Just replace the day of the week choice with the start time. There is no error as end time can be less then start time after midnight. Perhaps a few moto-cross users will have to record their Enduro's in parts but for most users this UI will be much more intuitive.
> 
> This is no where near that clever. Let's see what's possible with a more extensive fix but this one should be kept simple.

Fair enough after all this is what I asked @sub3 to change on the backend. 

The extensive fix probably needs to figure out if the use case for the End Date spinner on manual recordings is worth the annoyance to users and then it becomes a trivial fix.

**Comment by phunkyfish** [comment]
Jenkins build this please

**Comment by phunkyfish** [comment]
@ksooo how do you feel about this:

- If setting the start day or time makes the end time invalid then automatically change the end time to start time + previous duration (or the instant recording duration). Otherwise the user must change the end time before the start time can be modified.
- If setting the end time to something invalid we could prompt the user so they are aware the value is invalid and that any invalid value will be replaced by start time plus instant recording duration.

**Comment by phunkyfish** [comment]
I think this might be the best we can do. As the day selector is a spinner it's not possible to easily set it to a specific day.

So maybe the best approach is to show a toast if an invalid date is selected so at least the user will know they have made an invalid selection.

**Comment by ksooo** [comment]
> then automatically change the end time to start time + pre

Keep it simple. For some users, no matter what you do, your chosen automatism will always do the wrong thing. I swear. And they will file issues. ;-)

My POV: Let the users put in whatever they want, but when OK is pressed, evaluate (in overridden OnSettingChanging) and open a message box telling what is wrong, refuse to close. This is, how every web form works nowadays.

**Comment by phunkyfish** [comment]
OnSettingChanging is called for each spinner click or time entry. What I really want is something that is called before save but I’m not sure it’s possible.

Maybe just a dialog on save if something is wrong slowing the sensible default?

**Comment by ksooo** [comment]
> Unless there is a way to cancel the closing of the timer settings UI

 We need that, imo. Everything else is just a workaround. If it is currently not possible, we should implement this.

 I think, we are not in a hurry. Let's do it right, even if we need to write more new code for this than we thought. We've lived with this bug for years... no rush. let's do it right.

**Comment by ksooo** [comment]
@Montellese might be of help here. Ping.

**Comment by Montellese** [comment]
Can you provide me some more details? What are the involved setting types and UI controls being used? What do you want to do? Not changing the value when it doesn't make sense or automatically choose a different value when the new one doesn't make sense?

**Comment by phunkyfish** [comment]
> Can you provide me some more details? What are the involved setting types and UI controls being used? What do you want to do? Not changing the value when it doesn't make sense or automatically choose a different value when the new one doesn't make sense?

Of course. Here is a screenshot showing the controls.

Screenshot 2021-01-11 at 10 25 13

Quite simply all I would like to be able to do is allow the user to select whatever values they wish and once they click OK be able to validate the data. Then I could choose to accept the input or display a message showing the error and leave the user on the settings UI form so they can correct the values.

The complexity arises because the start and end times are split across multiple fields (Start date, start time, end date and end time). And you need all four to be able to test the validity.

**Comment by Montellese** [comment]
I think the best solution for your problem is to
* change the signature of `CGUIDialogSettingsBase::OnOkay()` to return a `bool` and change all usages to `return true;` at the end.
* adjust all (probably only one) usages of `OnOkay()` to only close the dialog if `OnOkay()` returns `true`.
* change the signature of `CGUIDialogSettingsManagerBase::Save()` to return a `bool` and change all usages to `return true;` at the end.
* adjust `CGUIDialogSettingsManagerBase::OnOkay()` to return `false` if `Save()` returns `false`.
* adjust the behaviour in `CGUIDialogPVRTimerSettings::Save()` to `false` if the timer settings are off.

**Comment by phunkyfish** [comment]
Cool, thanks. Let me take a look and see how much change it involves.

**Comment by emveepee** [comment]
Are you really sure you need to support > 24hr recordings?  For me  this is much easier for the user (commented out the End Date scroller for this shot) and you can use the auto day rollover that is there for the TimerRule.  

image

**Comment by phunkyfish** [comment]
> Are you really sure you need to support > 24hr recordings? For me this is much easier for the user (commented out the End Date scroller for this shot) and you can use the auto day rollover that is there for the TimerRule.
> 
> image

Right now I don’t want to restrict existing behaviour in any way. It’s too late in the day for that.

**Pull request merged**